src/main/java/org/cactoos/scalar/CheckedScalar.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2018 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.scalar;
+
+import org.cactoos.Func;
+import org.cactoos.Scalar;
+import org.cactoos.func.UncheckedFunc;
+
+/**
+ * Scalar that throws exception specified by user.
+ *
+ *  <p>There is no thread-safety guarantee.
+ *
+ * @author Vedran Vatavuk (123vgv@gmail.com)
+ * @version $Id$
+ * @param <T> Type of result
+ * @param <E> Type of exception
+ * @since 0.30
+ */
+public final class CheckedScalar<T, E extends Exception> implements Scalar<T> {
+
+    /**
+     * Function that wraps exception.
+     */
+    private final Func<Exception, E> func;
+
+    /**
+     * Original scalar.
+     */
+    private final Scalar<T> origin;
+
+    /**
+     * Ctor.
+     * @param scalar Encapsulated scalar
+     * @param func Func that wraps exception
+     */
+    public CheckedScalar(final Scalar<T> scalar,
+        final Func<Exception, E> func) {
+        this.func = func;
+        this.origin = scalar;
+    }
+
+    @Override
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    public T value() throws E {
+        try {
+            return this.origin.value();
+            // @checkstyle IllegalCatchCheck (1 line)
+        } catch (final RuntimeException ex) {
+            throw this.wrappedException(ex);
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw this.wrappedException(ex);
+            // @checkstyle IllegalCatchCheck (1 line)
+        } catch (final Exception ex) {
+            throw this.wrappedException(ex);
+        }
+    }
+
+    /**
+     * Wraps exception. Skips unnecessary boxing of exception.
+     * @param exp Exception
+     * @return E Wrapped exception
+     */
+    @SuppressWarnings("unchecked")
+    private E wrappedException(final Exception exp) {
```

Comment:
@Vatavuk I think the whole this logic about comparing exceptions is wrong and might lead to a problem: let's say I have this code
```
new CheckedScalar<Integer, IOException>(
    new ScalarOf(1),
    (ex) ->new IOException("New custom message", ex)
);
```
In case of IOException thrown in scalar, your logic will erase my error message, but I wouldn't expect this. So, I would suggest to not add this extra-logic and just throw what the function returns.

Reply:
@proshin-roman You are right. I  introduced that logic because I was concerned about unnecessary boxing of exceptions but above example clearly reveals a side effect. I removed it. 
I also change the variable name although there are many classes which use `func` as a field variable name. 
For example: `SolidScalar`, `StickyScalar`, `RetryScalar`...